File: frontend/src/lib/models/Rating.ts
import type { RatingResponse, RatingBuyerInfo } from '../types'

export class Rating {
  id: number
  store_id: number
  buyer_id: number
  order_id: number | null
  score: number
  comment: string | null
  created_at: string
  updated_at: string
  buyer: RatingBuyerInfo | null

  constructor(dto: RatingResponse) {
    this.id = dto.id
    this.store_id = dto.store_id
    this.buyer_id = dto.buyer_id
    this.order_id = dto.order_id ?? null
    this.score = dto.score
    this.comment = dto.comment ?? null
    this.created_at = dto.created_at
    this.updated_at = dto.updated_at
    this.buyer = dto.buyer ?? null
  }

  hasComment(): boolean {
    return this.comment !== null && this.comment.trim().length > 0
  }

  shortComment(maxLen: number = 50): string {
    if (!this.comment) return ''
    if (this.comment.length <= maxLen) return this.comment
    return this.comment.substring(0, maxLen) + '…'
  }

  static fromDto(dto: RatingResponse): Rating {
    return new Rating(dto)
  }
}

File: backend/app/schemas/rating.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from decimal import Decimal

class RatingCreate(BaseModel):
    store_id: int
    score: int = Field(..., ge=1, le=5)
    comment: Optional[str] = Field(None, max_length=1000)
    order_id: Optional[int] = None

class RatingUpdate(BaseModel):
    score: Optional[int] = Field(None, ge=1, le=5)
    comment: Optional[str] = Field(None, max_length=1000)

class RatingBuyerInfo(BaseModel):
    id: int
    username: str
    avatar_url: Optional[str] = None

    class Config:
        from_attributes = True

class RatingResponse(BaseModel):
    id: int
    store_id: int
    buyer_id: int
    order_id: Optional[int] = None
    score: int
    comment: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    buyer: Optional[RatingBuyerInfo] = None

    class Config:
        from_attributes = True

class StoreSummaryRating(BaseModel):
    store_id: int
    average_score: Optional[Decimal] = None
    total_ratings: int
    ratings: List[RatingResponse] = []